rustls/src/hash_hs.rs

```diff
@@ -129,11 +123,14 @@ impl HandshakeHash {
     /// 'handshake_hash' handshake message.  Start this hash
     /// again, with that message at the front.
     pub fn rollup_for_hrr(&mut self) {
-        let old_hash = self.ctx.take().unwrap().finish();
+        let ctx = self.ctx.as_mut().unwrap();
```

Comment:
I don't understand why this new version of `rollup_for_hrr` is better than the previous version. I think we should revert the changes to `rollup_for_hrr` and leave the rest.

It seems like this could stay as `take()` to simplify the code below.

Reply:
If we take().finish(), I think we no longer have access to the previous ctx to refer to its algorithm. I also prefer this version because it maintains the invariant of self.ctx.is_some() throughout the call.

Reply:
To be clear, you could `take()` and then get the algorithm and then `finish()`. However, now I understand better what you're trying to do, and I'm fine with it. We just need to remove the duplicate digest context creation.